8255918: XMLStreamFilterImpl constructor consumes XMLStreamException

MikeEdgar opened the pull request:
Allow `XMLStreamException` to be thrown to the application when a filtered `XMLStreamReader` encounters an `XMLStreamException` advancing the wrapped `XMLStreamReader`.

Note, this PR includes a method signature change (constructor of `com.sun.org.apache.xerces.internal.impl.XMLStreamFilterImpl`).

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8255918](https://bugs.openjdk.java.net/browse/JDK-8255918): XMLStreamFilterImpl constructor consumes XMLStreamException


### Reviewers
 * [Joe Wang](https://openjdk.java.net/census#joehw) (@JoeWang-Java - **Reviewer**) ⚠️ Review applies to a416b2768f272c29019b0b07d5c1862fbda5cd07


### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1209/head:pull/1209`
`$ git checkout pull/1209`

MikeEdgar commented:
/signed

MikeEdgar commented:
Test has been updated for `jtreg`.

JoeWang-Java commented on test/jdk/javax/xml/jaxp/parsers/8255918/XMLStreamReaderFilterTest.java:
```diff
@@ -0,0 +1,78 @@
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
```
Add package stream.XMLStreamFilterTest;

JoeWang-Java commented on test/jdk/javax/xml/jaxp/parsers/8255918/XMLStreamReaderFilterTest.java:
```diff
@@ -0,0 +1,78 @@
+        }
+
+        assertTrue(thrown instanceof XMLStreamException, "Missing or unexpected exception type: " + String.valueOf(thrown));
+        assertEquals(EXPECTED_MESSAGE1, thrown.getMessage(), "Unexpected exception message: " + thrown.getMessage());
```
It would be good/sufficient to verify XMLStreamException with assertThrows. Checking the message details can be fragile.

JoeWang-Java commented on test/jdk/javax/xml/jaxp/parsers/8255918/XMLStreamReaderFilterTest.java:
```diff
@@ -0,0 +1,78 @@
+     * @test
+     * @modules java.xml
+     * @run testng/othervm XMLStreamReaderFilterTest
+     *
```
Tags are added in the class description. Change this to a TestCase javadoc or note.

JoeWang-Java reviewed (commented):
I left comments the other day but forgot to submit. I've adjusted them based on your new changes.

JoeWang-Java commented on test/jdk/javax/xml/jaxp/parsers/8255918/XMLStreamReaderFilterTest.java:
```diff
@@ -0,0 +1,75 @@
+import javax.xml.stream.XMLStreamReader;
+
+/**
+ * JDK-8255918
```
This comment was shown under "static final String XMLSOURCE1" in the email, though in "Files changed" pane it still is right below the class description "JDK-8255918".

The comment meant to be class description, replacing "JDK-8255918".

MikeEdgar commented:
@JoeWang-Java - I believe I have addressed all of your comments. The branch has also been re-based on the `master` branch. Please let me know if any additional changes are needed.

JoeWang-Java commented on test/jaxp/javax/xml/jaxp/unittest/stream/XMLStreamFilterTest/XMLStreamReaderFilterTest.java:
```diff
@@ -0,0 +1,74 @@
+     * Test that construction of a filtered {@code XMLStreamReader} results in an
+     * {@code XMLStreamException} thrown when the underlying XML is not valid and
+     * the filter condition requires that the original reader is advanced past the
+     * invalid data.
```
This test verifies an assertion made by the javadoc for the createFilteredReader method. It would be good to made the first sentence a summary of the test. Something like:

Verifies that XMLStreamException is thrown as specified by the {@code XMLInputFactory::createFilteredReader} method when an error is encountered. This test illustrates the scenario by creating a reader with a filter that requires the original reader to advance past the invalid element in the underlying XML.

JoeWang-Java commented on test/jaxp/javax/xml/jaxp/unittest/stream/XMLStreamFilterTest/XMLStreamReaderFilterTest.java:
```diff
@@ -0,0 +1,74 @@
+     * @throws Exception When an unexpected exception is encountered (test failure)
+     */
+    @Test
+    public void testXMLStreamReaderExceptionThrownByConstructor() throws Exception {
```
Method name is a bit long. Since the class is dedicated to testing StreamFilter, its test cases may just indicate their particular cases. I would think testCreateFilteredReader is sufficient.

JoeWang-Java commented on test/jaxp/javax/xml/jaxp/unittest/stream/XMLStreamFilterTest/XMLStreamReaderFilterTest.java:
```diff
@@ -0,0 +1,74 @@
+     */
+    @Test
+    public void testXMLStreamReaderExceptionThrownByConstructor() throws Exception {
+        StreamFilter filter = r -> r.getEventType() == XMLStreamConstants.START_ELEMENT && r.getLocalName().equals("element3");
```
Keeping the length to under 100 characters is still a good practice even with wide screens, 80 - 100 would be good as it's easier to read with a side-by-side view.

JoeWang-Java reviewed (approved):
Looks good!  Some minor comments below.

MikeEdgar commented:
@JoeWang-Java - thanks for the additional feedback - all good. Do you suggest I make them now in this PR? Just confirming before pushing the change and dismissing your approval.

JoeWang-Java commented:
Yes, please make the changes. Approval will be quick :-) (since this is the last day for low priority changes)

MikeEdgar commented:
Done - please let me know if anything else is needed.

MikeEdgar commented:
/integrate

JoeWang-Java commented:
/sponsor
